urt/man/man1/rlequant.1
.\" Copyright (c) 1990, University of Michigan
.TH RLEQUANT 1 "June 12, 1990" 1
.UC 4 
.SH NAME
rlequant \- variance based color quantization for RLE images
.SH SYNOPSIS
.B rlequant
[
.B \-b
.I bits
] [
.B \-c
.I colors
] [
.B \-d
] [
.B \-f
] [
.B \-o
.I outfile
] [ 
.I infile
] 
.SH DESCRIPTION
This program quantizes the colors in an RLE image using a
variance-based method.  See 
.IR colorquant (3)
for more details on the method.
.TP
.BI \-b " bits"
The colors in the input image will be "prequantized" to this many bits
before applying the variance-based method.  Two internal tables of
size 
.BR "2^(3*" bits ")"
are allocated, so values of
.I bits
greater than 6 are likely to cause thrashing or may prevent the
program from running at all.  The default value of
.I bits
is 5.  It must be less than or equal to 8 and greater than 0.
.TP
.BI \-c " colors"
The output image will be quantized to at most
.I colors
colors.  It might have fewer if the input image has only a few colors
itself.  The default value of
.I colors
is 256.  It must be less than or equal to 256.
.TP
.B \-d
Floyd Steinberg dithering is performed on the output.  This is very helpful
for images being quantized to a small number of colors.
.TP
.B \-f
If this option is specified, a faster approximation will be used.  In
most cases, the error so introduced will be barely noticeable.
.TP
.BI \-o " outfile"
The output will be written to the file
.IR outfile .
If not specified, or if
.I outfile
is "\-", the output will be written to the standard output stream.
.TP
.I infile
This file contains one or more concatenated RLE images.  Each will be
processed in turn.  A separate quantization map will be constructed
for each image.  If not specified, or if
.I infile
is "\-", the image(s) will be read from the standard input stream.
.SH SEE ALSO
.IR mcut (1),
.IR rledither (1),
.IR urt (1),
.IR colorquant (3),
.IR RLE (5).
.SH AUTHOR
Spencer W. Thomas
.br
Craig Kolb (Yale University) wrote the color quantization code.
.br
Rod Bogart wrote the dithering code.
.SH BUGS
If you specify 
.BR \-d ,
it needs to compute a complete inverse color map.  This takes a long
time (especially if you don't specify
.BR \-f ).
That's life.  


urt/tools/rlequant.c
/*
 * This software is copyrighted as noted below.  It may be freely copied,
 * modified, and redistributed, provided that the copyright notice is 
 * preserved on all copies.
 * 
 * There is no warranty or other guarantee of fitness for this software,
 * it is provided solely "as is".  Bug reports or fixes may be sent
 * to the author, who may or may not act on them as he desires.
 *
 * You may not include this software in a program or other software product
 * without supplying the source, or without informing the end-user that the 
 * source is available for no extra charge.
 *
 * If you modify this software, you should include a notice giving the
 * name of the person performing the modification, the date of modification,
 * and the reason for such modification.
 */
/* 
 * rlequant.c - Quantize an image to a given number of colors.
 * 
 * Author:	Spencer W. Thomas
 * 		EECS Dept.
 * 		University of Michigan
 * Date:	Tue Jun 12 1990
 * Copyright (c) 1990, University of Michigan
 */

#include <stdio.h>
#include <rle.h>

#ifdef USE_STDLIB_H
#include <stdlib.h>
#else

#ifdef VOID_STAR
extern void *malloc();
#else
extern char *malloc();
#endif
extern void free();

#endif /* USE_STDLIB_H */

#define Quantize(x)	(x >> shift)

static void mem_alloc(), read_input();
static void setup_output(), write_output(), free_mem();

static CONST_DECL char *MY_NAME = "rlequant";

/*****************************************************************
 * TAG( main )
 *
 * Using Craig Kolb's colorquant code, quantize an RLE image to a
 * specified number of colors.
 * Usage:
 * 	rlequant [-b bits] [-c colors] [-d] [-f] [-o outfile] [infile]
 * Inputs:
 * 	-b bits:	The number of bits to which the image will be
 * 			"prequantized".  The default is 5.  Normally
 * 			this is fine, but some images have a very
 * 			limited number of colors.  The program will
 * 			require more memory if a larger value is
 * 			given: an array of size 2^(3*bits) is
 * 			allocated.  Must be <= 8.
 * 	-c colors:	The number of colors to quantize the image to.
 * 			The output image will contain no more than
 * 			this number of colors.  Must be <= 256.
 *	-d:		Floyd Steinberg Dither the output.
 * 	-f:		"Fast" mode.  The quantization accuracy will
 * 			be a little worse, but it is usually acceptable.
 * 	infile:		The input RLE file.  Default stdin.
 * 			"-" means stdin.
 * Outputs:
 * 	-o outfile:	The output RLE file.  Default stdout.
 * 			"-" means stdout.  The output image will be a
 * 			single channel image with a color map.
 * Assumptions:
 *	[None]
 * Algorithm:
 * 	Read image, call colorquant, write image with new colormap.
 */
void
main( argc, argv )
int argc;
char **argv;
{
    char       *infname = NULL,
    	       *outfname = NULL;
    int 	oflag = 0;
    int		bflag = 0,
    		bits = 5;
    int		cflag = 0,
    		colors = 256;
    int		fflag = 0;
    int		dflag = 0;
    int		zero = 0;
    int		rle_cnt, rle_err, width, height, shift;
    long	entries, bytecount;
    FILE       *outfile = stdout;
    rle_hdr in_hdr, out_hdr;	/* Headers for input and output files. */
    rle_pixel **rows;		/* Will be used for scanline storage. */
    rle_pixel *outrows[2];	/* For the output scanlines. */
    rle_pixel *alpha = NULL, *red, *green, *blue;	/* Image storage. */
    rle_pixel *img_red, *img_green, *img_blue; /* Image storage. */
    rle_pixel *colormap[3];	/* The quantized colormap. */
    rle_pixel *rgbmap;		/* Used for quantization. */

    MY_NAME = cmd_name( argv );

    if ( scanargs( argc, argv,
		   "% b%-bits!d c%-colors!d d%- f%- o%-outfile!s infile%s",
		   &bflag, &bits, &cflag, &colors, &dflag, &fflag,
		   &oflag, &outfname, &infname ) == 0 )
	exit( 1 );

    if ( bits <= 0 || bits > 8 )
    {
	fprintf( stderr, "%s: The bits argument must be >0 and <= 8.\n",
		 MY_NAME );
	exit( 1 );
    }
    if ( colors <= 0 || colors > 256 )
    {
	fprintf( stderr, "%s: The colors argument must be >0 and <= 256.\n",
		 MY_NAME );
	exit( 1 );
    }

    /* Open the input file.
     * The output file won't be opened until the first image header
     * has been read.  This avoids unnecessarily wiping out a
     * pre-existing file if the input is garbage.
     */
    in_hdr.rle_file = rle_open_f( MY_NAME, infname, "r" );

    /* Allocate the new colormap. */
    colormap[0] = (rle_pixel *)malloc( colors );
    colormap[1] = (rle_pixel *)malloc( colors );
    colormap[2] = (rle_pixel *)malloc( colors );

    /* bytecount gives the size of the rgbmap, and is equal to (2^bits)^3
     * Allocate memory for the rgbmap.
     */
    bytecount = 1 << bits;
    bytecount *= bytecount * bytecount;
    rgbmap = (rle_pixel *)malloc( bytecount );
    if ( rgbmap == NULL )
    {
	fprintf( stderr, "%s: Unable to allocate RGB map memory.\n",
		 MY_NAME );
	exit( RLE_NO_SPACE );
    }

    /* Amount to shift incoming pixels by for prequantization. */
    shift = 8 - bits;

    /* Read images from the input file until the end of file is
     * encountered or an error occurs.
     */
    rle_cnt = 0;
    while ( (rle_err = rle_get_setup( &in_hdr )) == RLE_SUCCESS )
    {
	/* Open the output file when the first header is successfully read. */
	if ( rle_cnt == 0 )
	    outfile = rle_open_f( MY_NAME, outfname, "w" );

	/* Count the input images. */
	rle_cnt++;

	/* Verify that the input image has at least 3 color channels. */
	if ( in_hdr.ncolors < 3 )
	{
	    fprintf( stderr, "%s: Input image %d has only %d colors.\n",
		     MY_NAME, rle_cnt, in_hdr.ncolors );
	    exit ( 1 );
	}

	/* The output header starts out as a copy of the input header.
	 * The FILE pointer is different, of course.
	 * Also, this output image will have only one channel.
	 */
	out_hdr = in_hdr;
	out_hdr.rle_file = outfile;
	out_hdr.ncolors = 1;
	out_hdr.bg_color = &zero;
	out_hdr.background = 2;

	rle_addhist( argv, &in_hdr, &out_hdr );

	/* Since rle_getrow and rle_putrow use different array origins,
	 * we will compensate by adjusting the xmin and xmax values in
	 * the input header.  [rle_getrow assumes that the scanline
	 * array starts at pixel 0, while rle_putrow assumes that the
	 * scanline array starts at pixel xmin.  This is a botch, but
	 * it's too late to change it now.]
	 */
	in_hdr.xmax -= in_hdr.xmin;
	in_hdr.xmin = 0;
	width = in_hdr.xmax + 1;	/* Width of a scanline. */

	/* For convenience, do the same to the Y values. */
	in_hdr.ymax -= in_hdr.ymin;
	in_hdr.ymin = 0;
	height = in_hdr.ymax + 1;	/* Height of image. */

	/* Total number of pixels in image. */
	entries = width * height;

	/* Allocate memory for the imput image. */
	mem_alloc( &in_hdr, entries, dflag, &rows, &red, &green, &blue, &alpha,
		   &img_red, &img_green, &img_blue );

	/* Read and quantize the input image. */
	read_input( &in_hdr, width, shift, dflag, red, green, blue, alpha,
		    img_red, img_green, img_blue, rows );

      /* Compute the color quantization map.  Dither flag implies that the
       * RGBmap must be built for multiple images.
       */
	colors = colorquant( red, green, blue, entries, colormap, colors, bits,
                           rgbmap, fflag, dflag );

	/* Compute and write setup information to the output file. */
	setup_output( &out_hdr, width, colors, colormap, outrows );

	/* Quantize (and maybe dither) the input image and write the
	 * output image.
	 */
	write_output( &out_hdr, width, height, bits, dflag,
		      red, green, blue, alpha,
		      img_red, img_green, img_blue, rgbmap, colormap,
		      outrows );


	/* Free all the memory we allocated. */
	free_mem( &in_hdr, dflag, red, green, blue, alpha,
		  img_red, img_green, img_blue, rows, outrows );
    }
    
    /* Check for an error.  EOF or EMPTY is ok if at least one image
     * has been read.  Otherwise, print an error message.
     */
    if ( rle_cnt == 0 || (rle_err != RLE_EOF && rle_err != RLE_EMPTY) )
	rle_get_error( rle_err, MY_NAME, infname );

    exit( 0 );
}


static void
mem_alloc( hdr, entries, dflag, rows, red, green, blue, alpha,
	   img_red, img_green, img_blue )
rle_hdr *hdr;
unsigned long entries;
int dflag;
rle_pixel ***rows;
rle_pixel **red, **green, **blue, **alpha;
rle_pixel **img_red, **img_green, **img_blue;
{
    /* Allocate memory into which the image scanlines can be read.
     * This should happen after the above adjustment, to minimize
     * the amount of memory allocated.
     */
    if ( rle_row_alloc( hdr, rows ) < 0 )
    {
	fprintf( stderr, "%s: Unable to allocate scanline memory.\n",
		 MY_NAME );
	exit( RLE_NO_SPACE );
    }
	
    /* Allocate image memory for prequantized image. */
    *red = (rle_pixel *)malloc( entries );
    *green = (rle_pixel *)malloc( entries );
    *blue = (rle_pixel *)malloc( entries );
    if ( hdr->alpha )
	*alpha = (rle_pixel *)malloc( entries );

    if ( (hdr->alpha && (*alpha == NULL)) || 
	 *red == NULL || *green == NULL || *blue == NULL )
    {
	fprintf( stderr, "%s: Unable to allocate image memory.\n",
		 MY_NAME );
	exit( RLE_NO_SPACE );
    }

    if (dflag) {
	/* Allocate image memory for original image to dither on. */
	*img_red = (rle_pixel *)malloc( entries );
	*img_green = (rle_pixel *)malloc( entries );
	*img_blue = (rle_pixel *)malloc( entries );

	if ( *img_red == NULL || *img_green == NULL || *img_blue == NULL )
	{
	    fprintf( stderr, "%s: Unable to allocate image memory.\n",
		     MY_NAME );
	    exit( RLE_NO_SPACE );
	}
    }
}

static void
read_input( hdr, width, shift, dflag, red, green, blue, alpha,
	    img_red, img_green, img_blue, rows )
rle_hdr *hdr;
int width, shift, dflag;
rle_pixel *red, *green, *blue, *alpha;
rle_pixel *img_red, *img_green, *img_blue;
rle_pixel **rows;
{
    register rle_pixel *rp, *gp, *bp, *ap;
    register rle_pixel *irp = NULL, *igp = NULL, *ibp = NULL;
    int x, y;

    /* Set traversal pointers. */
    rp = red;
    gp = green;
    bp = blue;
    ap = alpha;
    if (dflag) {
	irp = img_red;
	igp = img_green;
	ibp = img_blue;
    }

    /* Read the input image and copy it to the output file. */
    for ( y = hdr->ymin; y <= hdr->ymax; y++ )
    {
	/* Read a scanline. */
	rle_getrow( hdr, rows );
	    
	/* Prequantize the pixels. */
	for ( x = 0; x < width; x++ )
	{
	    *rp++ = Quantize( rows[0][x] );
	    *gp++ = Quantize( rows[1][x] );
	    *bp++ = Quantize( rows[2][x] );
	    if ( hdr->alpha )
		*ap++ = rows[-1][x];
	    if (dflag) {
		*irp++ = rows[0][x];
		*igp++ = rows[1][x];
		*ibp++ = rows[2][x];
	    }
	}
    }
}

static void
setup_output( hdr, width, colors, colormap, outrows )
rle_hdr *hdr;
int colors, width;
rle_pixel *colormap[3];
rle_pixel *outrows[2];
{
    char *buf = (char *)malloc( 80 );
    int x, y;

    /* Give the output image a colormap. */
    hdr->cmap = (rle_map *)malloc( 3 * 256 * sizeof( rle_map ) );
    hdr->ncmap = 3;
    hdr->cmaplen = 8;
    for ( y = 0; y < 3; y++ )
    {
	for ( x = 0; x < colors; x++ )
	    hdr->cmap[y * 256 + x] = colormap[y][x] << 8;
	for ( ; x < 256; x++ )
	    hdr->cmap[y * 256 + x] = 0;
    }

    /* Add a comment to the output image with the true colormap length. */
    sprintf( buf, "color_map_length=%d", colors );
    rle_putcom( buf, hdr );

    /* Write the output image header. */
    rle_put_setup( hdr );

    /* Allocate memory for the output scanline. */
    outrows[1] = (rle_pixel *)malloc( width );
    if ( outrows[1] == NULL )
    {
	fprintf( stderr, "%s: Unable to allocate scanline memory.\n",
		 MY_NAME );
	exit( RLE_NO_SPACE );
    }

}

static void
write_output( hdr, width, height, bits, dflag, red, green, blue, alpha,
	      img_red, img_green, img_blue, rgbmap, colormap, outrows )
rle_hdr *hdr;
int width, height, bits, dflag;
rle_pixel *red, *green, *blue, *alpha;
rle_pixel *img_red, *img_green, *img_blue;
rle_pixel *rgbmap;
rle_pixel *colormap[3];
rle_pixel *outrows[2];
{
    int shift = 8 - bits;
    register rle_pixel *rp, *gp, *bp, *ap;
    register rle_pixel *irp, *igp, *ibp;
    int x, y;

    ap = alpha;
    if (!dflag)  {
	/* Set traversal pointers. */
	rp = red;
	gp = green;
	bp = blue;

	/* Write the output image. */
	for ( y = 0; y < height; y++ )
	{
	    /* Point to the correct data. */
	    if ( hdr->alpha )
	    {
		outrows[0] = ap;
		ap += width;
	    }
	    for ( x = 0; x < width; x++, rp++, gp++, bp++ )
		outrows[1][x] = rgbmap[(((*rp<<bits)|*gp)<<bits)|*bp];

	    rle_putrow( &outrows[1], width, hdr );
	}
    }
    else {
	register short *thisptr, *nextptr ;
	int	lastline, lastpixel ;
	short *thisline, *nextline, *tmpptr;

	/* Set traversal pointers. */
	irp = img_red;
	igp = img_green;
	ibp = img_blue;

	thisline = (short *) malloc( width * 3 * sizeof(short));
	nextline = (short *) malloc( width * 3 * sizeof(short));
	nextptr = nextline;
	for (x=0; x < width; x++)
	{
	    *nextptr++ = *irp++ ;
	    *nextptr++ = *igp++ ;
	    *nextptr++ = *ibp++ ;
	}
	for (y=0; y < height; y++)
	{
	    /* swap nextline into thisline and copy new nextline */
	    tmpptr = thisline ;
	    thisline = nextline ;
	    nextline = tmpptr ;
	    lastline = (y == height - 1) ;
	    if (!lastline)
	    {
		nextptr = nextline;
		for (x=0; x < width; x++)
		{
		    *nextptr++ = *irp++ ;
		    *nextptr++ = *igp++ ;
		    *nextptr++ = *ibp++ ;
		}
	    }

	    thisptr = thisline ;
	    nextptr = nextline ;

	    for(x=0; x < width ; x++)
	    {
		int	rval, gval, bval, color ;
		rle_pixel r2,g2,b2 ;

		lastpixel = (x == width - 1) ;

		rval = *thisptr++ ;
		gval = *thisptr++ ;
		bval = *thisptr++ ;

		/* Current pixel has been accumulating error, it could be
		 * out of range.
		 */
		if( rval < 0 ) rval = 0 ;
		else if( rval > 255 ) rval = 255 ;
		if( gval < 0 ) gval = 0 ;
		else if( gval > 255 ) gval = 255 ;
		if( bval < 0 ) bval = 0 ;
		else if( bval > 255 ) bval = 255 ;

		r2 = Quantize(rval);
		g2 = Quantize(gval);
		b2 = Quantize(bval);

		color = rgbmap[(((r2<<bits)|g2)<<bits)|b2];
		outrows[1][x] = color;

		rval -= colormap[0][color];
		gval -= colormap[1][color];
		bval -= colormap[2][color];

		if( !lastpixel )
		{
		    thisptr[0] += rval * 7 / 16 ;
		    thisptr[1] += gval * 7 / 16 ;
		    thisptr[2] += bval * 7 / 16 ;
		}
		if( !lastline )
		{
		    if( x != 0 )
		    {
			nextptr[-3] += rval * 3 / 16 ;
			nextptr[-2] += gval * 3 / 16 ;
			nextptr[-1] += bval * 3 / 16 ;
		    }
		    nextptr[0] += rval * 5 / 16 ;
		    nextptr[1] += gval * 5 / 16 ;
		    nextptr[2] += bval * 5 / 16 ;
		    if( !lastpixel )
		    {
			nextptr[3] += rval / 16 ;
			nextptr[4] += gval / 16 ;
			nextptr[5] += bval / 16 ;
		    }
		    nextptr += 3 ;
		}
	    }

	    if ( hdr->alpha )
	    {
		outrows[0] = ap;
		ap += width;
	    }
	    rle_putrow( &outrows[1], width, hdr );
	}
    }

    /* Write an end-of-image code. */
    rle_puteof( hdr );
}


static void
free_mem( hdr, dflag, red, green, blue, alpha,
	  img_red, img_green, img_blue, rows, outrows )
rle_hdr *hdr;
int dflag;
rle_pixel *red, *green, *blue, *alpha;
rle_pixel *img_red, *img_green, *img_blue;
rle_pixel **rows;
rle_pixel *outrows[2];
{
    /* Free image store. */
    free( red );
    free( green );
    free( blue );
    if ( hdr->alpha )
	free( alpha );
    if (dflag) {
	/* Free image store. */
	free( img_red );
	free( img_green );
	free( img_blue );
    }
    /* Free the scanline memory. */
    rle_row_free( hdr, rows );
    free( outrows[1] );
}
